src/mono/mono/mini/cpu-amd64.mdesc

```diff
@@ -843,6 +843,7 @@ ssse3_shuffle: dest:x src1:x src2:x len:6 clob:1
 sse41_dpps_imm: dest:x src1:x src2:x len:7 clob:1
 sse41_dppd_imm: dest:x src1:x src2:x len:7 clob:1
 vector_andnot: dest:x src1:x src2:x len:7 clob:1
+vector_integer_abs: dest:x src1:x len:16
```

Comment:
Why is this `len:16`? Shouldn't it in the worst case be `len:8` since its a 3-byte `movups` + the 5-byte `pabs` instruction?

Reply:
I did put 16 as I did not know what value to put there and was advised that it should be sufficient to put a large number there for now. Could you share how to calculate the max len correctly?

I think it doesn't need to specify `clob:1` as it only writes to the destination register

Reply:
The max length is based on the instruction encoding, its only used to calculate x86 short branches etc. so its ok to overstate it, esp. for rarely used opcodes. Here, the encoding is emit_sse_reg_reg_op4, which has a length of 6.

Reply:
So in case of emit_sse_reg_reg_op4 it is 4 bytes for  instruction and 1 byte per register?

Reply:
@vargaz Could you explain how it's 6?

For something like `vector_andnot` it's specified as `7` today:
* You have a 3 byte opcode (0x66, 0x0F, 0xDF) + the modr/m byte, giving it a 4-byte base encoding
* If you use an extended register that's +1 additional for the prefix
* If you use an extended a non-basic addressing mode that's +1 additional for the SIB byte
* If you use a displacement that's +4 bytes.

So the actual encoding is 4-10 bytes, depending on what can or can't be included. Since its RMW (the destination is also an input operand) you may also need to emit a `movups`/`movaps`/`movdqu` and that may or may not need to be included.

So, I'd expect `vector_andnot` to be specified as one of:
* `4` (base encoding)
* `5` (base encoding + extended register) -or- (base encoding + complex addressing mode)
* `6` (base encoding + extended register + complex addressing mode)
* `9` (base encoding + complex addressing mode + displacement)
* `10` (base encoding + extended register + complex addressing mode + displacement)
  * potentially higher if it also factors in RMW handling
  * none of this factors in using the VEX encoding, which I don't believe is supported by Mono today

So I'd expect `pabs*` to follow a similar model, whatever that is above, and effectively be `1` more than `vector_andnot` (since it has a 5 byte base encoding), and only be lesser if RMW handling is factored into the length, since `pabs` doesn't need it.

The `clob:1` setting looks to indicate it clobbers the first source register, which I don't follow how that works since the `dest == source` for RMW instructions (when not using the VEX encoding) and you never have to clobber it, it should always have its original value preserved unless its last use, in which case clobbering doesn't matter.

Reply:
Its emitted using this macro:

#define emit_sse_reg_reg_op4_size(inst,dreg,reg,op1,op2,op3,op4,size) do { \
    amd64_codegen_pre(inst); \
    x86_prefix((inst), (unsigned char)(op1)); \
    amd64_emit_rex ((inst), size, (dreg), 0, (reg)); \
    *(inst)++ = (unsigned char)(op2); \
    *(inst)++ = (unsigned char)(op3); \
    *(inst)++ = (unsigned char)(op4); \
    x86_reg_emit ((inst), (dreg), (reg)); \
    amd64_codegen_post(inst); \
} while (0)

Which afaiks is 6 bytes.

clob:1 is used for the 2 operand x86 binary operation instructions where dreg needs to be equal to sreg1. It tells the register allocator to allocate both to the same hw register. So it doesn't really clobber anything.